llvm/include/llvm/Target/GlobalISel/Combine.td

```diff
@@ -323,6 +323,24 @@ def reduce_shl_of_extend : GICombineRule<
          [{ return Helper.matchCombineShlOfExtend(*${mi}, ${matchinfo}); }]),
   (apply [{ Helper.applyCombineShlOfExtend(*${mi}, ${matchinfo}); }])>;
 
+// Combine bitreverse(shl (bitreverse x), y)) -> (lshr x, y)
+def bitreverse_shl : GICombineRule<
+  (defs root:$d, build_fn_matchinfo:$matchinfo),
+  (match (G_BITREVERSE $rev, $val),
+         (G_SHL $src, $rev, $amt),
+         (G_BITREVERSE $d, $src):$mi,
+         [{ return Helper.matchBitreverseShift(*${mi}, ${matchinfo}); }]),
+  (apply [{ Helper.applyBuildFn(*${mi}, ${matchinfo}); }])>;
+
+// Combine bitreverse(lshr (bitreverse x), y)) -> (shl x, y)
+def bitreverse_lshr : GICombineRule<
+  (defs root:$d, build_fn_matchinfo:$matchinfo),
+  (match (G_BITREVERSE $rev, $val),
+         (G_LSHR $src, $rev, $amt),
+         (G_BITREVERSE $d, $src):$mi,
+         [{ return Helper.matchBitreverseShift(*${mi}, ${matchinfo}); }]),
+  (apply [{ Helper.applyBuildFn(*${mi}, ${matchinfo}); }])>;
```

Comment:
I've had to keep the matchBitreverseShift for legality tests, but the full patterns are now tested in the td file.

Reply:
Good job. If you match for G_BITREVERSE, then the combiner will jump for every G_BITREVERSE into C++ and fail in most cases.

Reply:
Is there anyway that we could replace the matchBitreverseShift call with a legality check?